data/game_settings.txt
{
    "user": 
    {
        // 头像id
        "user_icons": [
            "icon:1000", "icon:1001", "icon:1002", "icon:1003", "icon:1004", 
            "icon:1005", "icon:1006", "icon:1007", "icon:1008", "icon:1009", 
            "icon:1010"
        ],
        
        // 用户等级经验
        "using_user_level_experiences_table": true,
        "user_level_experiences": [
            0,     100,   300,   500,   1500,  2000,  2500,  3000,  3500,  4000, 
            4500,  5000,  6000,  7000,  8000,  9000,  10000, 11000, 12000, 13000, 
            13000, 15000, 18000, 21000, 24000, 28000, 32000, 36000, 41000, 46000,
            51000
        ],

        //
        "item_default_equipment": 101,

        //
        // 改名必须消耗的物品，目前只有改名卡
        "need_change_name_items": "5010|1",
        // 7天才能修改 604800秒
        "need_change_name_time": 604800

    },

    "vip":
    {
        "enabled": true,
        "code": 100001,
        "experience_ratio": 10.0
    },

    "season":
    {
        "index": 1,
        "code": 1001,
        "pass": {
            "0":[ ],
            "1":[ 101000 ],
            "2":[ 101001 ]
        }
    },

    "leaderboard":
    {
        "gold_limit_min": 1000000,
        "gems_limit_min": 5000,
        "gems_limit_min_weekly": 10000,
        "gems_limit_min_monthly": 10000,
        // 上榜最小段位
        "game_rank_min_level": 1001
    },

    // 每日任务
    "mission":
    {
        "daily_enabled": true,
        "daily_event_code": 1000
    }
}

data/payment_settings.txt
{
    "ssl_certs": "certs/payment.crt",
    "ssl_key": "certs/payment.pem.unsecure",
    // 订单超时时间，一分钟内未完成支付，按超时处理。
    "transaction_timeout": 300,
    "wechat": {
        "enabled": true,
        "is_sandbox": true,
        "sandbox_url": "https://api.mch.weixin.qq.com",
        "url": "https://api.mch.weixin.qq.com",
        "sandbox_app_id": "wx9fa8acf808811f3f",
        "app_id": "wx9fa8acf808811f3f",
        // 方法
        "method": "",
        // 商户ID
        "merchant_id": "123456",
        "notify_url": "https://47.98.229.74:5443/openapi/payment/final"
    },
    "alipay": {
        "enabled": true,
        "is_sandbox": true,
        "sandbox_url": "https://openapi-sandbox.dl.alipaydev.com/gateway.do",
        "url": "https://openapi.alipay.com/gateway.do",
        "sandbox_app_id": "9021000152623602",
        "app_id": "2021005187659435",
        // 方法
        "method": "alipay.trade.query"
    }
}


src/Config.cs

using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace Server {

    public interface IConfigEntry
    {

    }

    /// <summary>
    /// 
    /// </summary>
    public class ConfigItem_Logger
    {
        [YamlMember(Alias = "enabled")]
        public bool Enabled { get; set; } = true; // 默认路径
        [YamlMember(Alias = "name")] // 明确映射 YAML 字段名（可选，如果字段名不一致时需要）
        public string Name { get; set; } = ""; // 默认值
        [YamlMember(Alias = "level")] // 明确映射 YAML 字段名（可选，如果字段名不一致时需要）
        public string Level { get; set; } = "Information"; // 默认值

        [YamlMember(Alias = "file")]
        public string File { get; set; } = "logs/main.log"; // 默认路径
        [YamlMember(Alias = "is_console", ApplyNamingConventions = false)]
        public bool IsConsole { get; set; } = true; // 默认开启
        [YamlMember(Alias = "is_file", ApplyNamingConventions = false)]
        public bool IsFile { get; set; } = true; // 默认开启
        public Logger.LogLevel Getlevel() {
            //
            if(Level == Logger.LogLevel.Debug.ToString())
            {
                return Logger.LogLevel.Debug;
            }
            else if(Level == Logger.LogLevel.Warning.ToString())
            {
                return Logger.LogLevel.Warning;
            }
            else if(Level == Logger.LogLevel.Error.ToString())
            {
                return Logger.LogLevel.Error;
            }
            else
            {
                return Logger.LogLevel.Information;
            }
        }
    }

    public class ConfigItem_HTTPServer
    {
        [YamlMember(Alias = "address")]
        public string Address { get; set; } = "0.0.0.0";
        [YamlMember(Alias = "port")]
        public int Port { get; set; } = 5000;


        [YamlMember(Alias = "ssl")]
        public bool HasSSL { get; set; } = false;
        [YamlMember(Alias = "ssl_certs", ApplyNamingConventions = false)]
        public string SSLCertificates { get; set; } = "";
        [YamlMember(Alias = "ssl_key", ApplyNamingConventions = false)]
        public string SSLKey { get; set; } = "";


        [YamlMember(Alias = "internal")]
        public bool IsInternalService { get; set; } = false;
        /// <summary>
        /// 仅仅只有允许访问的IP可以访问
        /// </summary>
        [YamlMember(Alias = "allow_address_list")]
        public string[] AllowAddressList { get; set; } = new string[] { "127.0.0.1" };
    }

    public class ConfigItem_WSServer
    {
        [YamlMember(Alias = "address")]
        public string Address { get; set; } = "0.0.0.0"; 
        [YamlMember(Alias = "port")]
        public int Port { get; set; } = 5900; 
        [YamlMember(Alias = "ssl")]
        public bool HasSSL { get; set; } = false;
        [YamlMember(Alias = "ssl_certs", ApplyNamingConventions = false)]
        public string SSLCertificates { get; set; } = "";
        [YamlMember(Alias = "ssl_key", ApplyNamingConventions = false)]
        public string SSLKey { get; set; } = "";
    }
    
    public class ConfigItem_Database
    {
        [YamlMember(Alias = "type")]
        public string Type { get; set; } = "main"; 
        [YamlMember(Alias = "name")]
        public string Name { get; set; } = "game"; 
        [YamlMember(Alias = "user")]
        public string UserName { get; set; } = "sa";
        [YamlMember(Alias = "pass")]
        public string Password { get; set; } = "123456";
        [YamlMember(Alias = "address")]
        public string Address { get; set; } = "127.0.0.1"; 
        [YamlMember(Alias = "port")]
        public int Port { get; set; } = 3306; 
        [YamlMember(Alias = "ssl")]
        public bool HasSSL { get; set; } = false;
        [YamlMember(Alias = "ssl_certs", ApplyNamingConventions = false)]
        public string SSLCertificates { get; set; } = "";
        [YamlMember(Alias = "ssl_key", ApplyNamingConventions = false)]
        public string SSLKey { get; set; } = "";
    }
    
    public class ConfigItem_Redis
    {
        [YamlMember(Alias = "name")]
        public string Name { get; set; } = "redis"; 
        [YamlMember(Alias = "user")]
        public string UserName { get; set; } = "sa";
        [YamlMember(Alias = "pass")]
        public string Password { get; set; } = "123456";
        [YamlMember(Alias = "address")]
        public string Address { get; set; } = "127.0.0.1"; 
        [YamlMember(Alias = "port")]
        public int Port { get; set; } = 6379; 
        [YamlMember(Alias = "ssl")]
        public bool HasSSL { get; set; } = false;
        [YamlMember(Alias = "ssl_certs", ApplyNamingConventions = false)]
        public string SSLCertificates { get; set; } = "";
        [YamlMember(Alias = "ssl_key", ApplyNamingConventions = false)]
        public string SSLKey { get; set; } = "";
    }

    public class ConfigItem_RoomManager
    {
        [YamlMember(Alias = "enabled")]
        public bool Enabled { get; set; } = true;
        [YamlMember(Alias = "room_max")]
        public int RoomMaxNum { get; set; } = 100;
        [YamlMember(Alias = "players_max")]
        public int PlayersMaxNum { get; set; } = 2;
        [YamlMember(Alias = "lookon_max")]
        public int LookOnMaxNum { get; set; } = 0;
    }

    public class ConfigItem_MatchServer
    {
        [YamlMember(Alias = "enabled")]
        public bool Enabled { get; set; } = true; 
        [YamlMember(Alias = "poll_max")]
        public int PollMaxNum { get; set; } = 100; 


        [YamlMember(Alias = "ai_player_enabled")]
        public bool IsAIPlayerEnabled { get; set; } = true;
        [YamlMember(Alias = "ai_player_max")]
        public int AIPlayerMaxNum { get; set; } = 1000; 
        [YamlMember(Alias = "ai_player_derived")]
        public bool IsAIPlayerDerived { get; set; } = true;

    }

    /// <summary>
    /// 
    /// </summary>
    public class ServerConfig : IConfigEntry
    {
        /// <summary>
        /// 
        /// </summary>
        [YamlMember(Alias = "server_index", ApplyNamingConventions = false)]
        public int ServerIndex { get; set; } = 0;
        
        [YamlMember(Alias = "secret_key", ApplyNamingConventions = false)]
        public string SecretKey { get; set; } = "";
        [YamlMember(Alias = "database_auth_enabled", ApplyNamingConventions = false)]
        public bool DBAuthorizationEnabled { get; set; } = false;
        [YamlMember(Alias = "database_auth_expired", ApplyNamingConventions = false)]
        public int DBAuthorizationExpired { get; set; } = 1 * 24 * 60 * 60;
        /// <summary>
        /// JWT 
        /// </summary>
        [YamlMember(Alias = "jwt_enabled", ApplyNamingConventions = false)]
        public bool JWTEnabled { get; set; } = true;
        [YamlMember(Alias = "jwt_auth_enabled", ApplyNamingConventions = false)]
        public bool JWTAuthorizationEnabled { get; set; } = true;
        /// <summary>
        /// JWT
        /// </summary>
        [YamlMember(Alias = "jwt_secret_key", ApplyNamingConventions = false)]
        public string JWTSecretKey { get; set; } = "";
        /// <summary>
        /// JWT Expired
        /// </summary>
        [YamlMember(Alias = "jwt_expired", ApplyNamingConventions = false)]
        public int JWTExpired { get; set; } = 1 * 24 * 60 * 60;
        /// <summary>
        /// 
        /// </summary>
        [YamlMember(Alias = "room", ApplyNamingConventions = false)]
        public ConfigItem_RoomManager Room = new ConfigItem_RoomManager();
        /// <summary>
        /// 
        /// </summary>
        [YamlMember(Alias = "match_server", ApplyNamingConventions = false)]
        public ConfigItem_MatchServer MatchServer = new ConfigItem_MatchServer();
        /// <summary>
        /// 
        /// </summary>
        [YamlMember(Alias = "server", ApplyNamingConventions = false)]
        public ConfigItem_WSServer WSServer { get; set; } = new ConfigItem_WSServer(); // 初始化默认配置
        /// <summary>
        /// 
        /// </summary>
        [YamlMember(Alias = "http_server", ApplyNamingConventions = false)]
        public ConfigItem_HTTPServer[] HTTPServer { get; set; } = new ConfigItem_HTTPServer[] { }; // 初始化默认配置

        /// <summary>
        /// PlayFab 第三方配置，不在此处定义
        /// </summary>
        [YamlMember(Alias = "playfab", ApplyNamingConventions = false)]
        public ConfigItem_PlayFab PlayFab { get; set; } = new ConfigItem_PlayFab (); // 初始化默认配置


        /// <summary>
        /// 
        /// </summary>
        [YamlMember(Alias = "database", ApplyNamingConventions = false)]
        public ConfigItem_Database[] Databases = new ConfigItem_Database[]{ };
        /// <summary>
        /// 
        /// </summary>
        [YamlMember(Alias = "redis", ApplyNamingConventions = false)]
        public ConfigItem_Redis Redis = new ConfigItem_Redis ();
        /// <summary>
        /// 
        /// </summary>
        [YamlMember(Alias = "logging", ApplyNamingConventions = false)]
        public ConfigItem_Logger[] Logging { get; set; } = new ConfigItem_Logger[]{
            new ConfigItem_Logger() {
                Name = "main",
                Level = "Information",
                File = "logs/main.log"
            }
        };
    }

    /// <summary>
    /// 
    /// </summary>
    public class ServerConfigLoader
    {
        private static ServerConfig _config = new ServerConfig();
        public static ServerConfig Config { get { return _config; } }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="filename"></param>
        /// <returns></returns>
        public static ServerConfig LoadFromFile(string filename)
        {
            try
            {
                // 读取 YAML 文件内容
                string yaml = File.ReadAllText(filename);


                // 创建反序列化器
                var deserializer = new DeserializerBuilder()
                    .WithNamingConvention(UnderscoredNamingConvention.Instance) // 自动下划线命名
                    //.WithNamingConvention(CamelCaseNamingConvention.Instance) // 自动驼峰命名
                    .Build();

                // 反序列化为对象
                var config = deserializer.Deserialize<ServerConfig>(yaml);
                if(config != null)
                {
                    _config = config;
                }

                _config.SecretKey = _config.SecretKey.Trim().ToUpper();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"加载配置文件失败: {ex.Message}");
            }
            return _config;
        }
    }
}
